### PR TITLE
Enhance Connection Collapse support in ATS core

### PR DESCRIPTION
**Pull request by sudheerv**
Add an option to support cache open read retry on a write lock
failure. With this option, as long as read-while-writer is set
up following the guidelines in the docs, there should be no need
for any plugins to augment the core. Eventual plan is to deprecate
collapsed_forwarding plugin with this new support.

For more context on this, see 
https://cwiki.apache.org/confluence/display/TS/Presentations+-+2019?preview=/112821251/132320653/Collapsed%20Forwarding%20.pdf

### CODE REVIEW

**Comment by scw00** [comment]
Does this increase the CACHE_LOOKUP_COMPLETE hooks calls ? 
What's the latency do we lost in these retries in worst  case ?

File: proxy/http/HttpSM.cc [hunk 1]
```diff
@@ -2451,6 +2451,12 @@ HttpSM::state_cache_open_write(int event, void *data)
   // INTENTIONAL FALL THROUGH
   // Allow for stale object to be served
   case CACHE_EVENT_OPEN_READ:
+    if (!t_state.cache_info.object_read) {
+      ink_assert(t_state.cache_open_write_fail_action == HttpTransact::CACHE_WL_FAIL_ACTION_READ_RETRY);
+      t_state.cache_lookup_result         = HttpTransact::CACHE_LOOKUP_NONE;
+      t_state.cache_info.write_lock_state = HttpTransact::CACHE_WL_READ_RETRY;
+      break;
```
Comment (scw00) [inline proxy/http/HttpSM.cc]: Why can `t_state.cache_info.object_read` be nullptr ?

Comment (scw00) [reply proxy/http/HttpSM.cc]: I see `state_cache_open_write ` call OPEN_READ to sm.

File: proxy/http/HttpTransact.cc [hunk 2]
```diff
@@ -2911,7 +2911,17 @@ HttpTransact::handle_cache_write_lock(State *s)
...
-    remove_ims                = true;
+    if (!s->cache_info.object_read) {
+      ink_assert(s->cache_open_write_fail_action == HttpTransact::CACHE_WL_FAIL_ACTION_READ_RETRY);
+      s->cache_info.write_status = CACHE_WRITE_LOCK_MISS;
+      StateMachineAction_t next;
+      next           = SM_ACTION_CACHE_LOOKUP;
+      s->next_action = next;
+      s->hdr_info.server_request.destroy();
+      TRANSACT_RETURN(next, nullptr);
+      return;
```
Comment (scw00) [inline proxy/http/HttpTransact.cc]: `return` can be removed

**Comment by scw00** [comment]
Please update:

https://github.com/apache/trafficserver/blob/12afa068a44a9400392256fdc64d16a891c326b5/proxy/http/HttpTransact.cc#L2906-L2909

**Comment by sudheerv** [comment]
> Hi Some question for this pr .
> 
> Does this increase the CACHE_LOOKUP_COMPLETE hook calls ?

No, the API callbacks with CACHE_LOOKUP_COMPLETE shouldn't be impacted. The call back for CACHE_LOOKUP_COMPLETE happens from HttpSM after HttpCacheSM finishes all retries and returns the final lookup status.

> What's the latency do we lost in these retries in worst case ?

The additional latency due to this new feature will be equal to (max_cache_open_read_retries*cache_open_read_retry_time) and the max extra latency will be equal to 2*(max_cache_open_read_retries*cache_open_read_retry_time).

Typical values used are max_cache_open_read_retries = 5-10, cache_open_read_retry_time = 10-50 msec, so a worst case of 1 sec (2 * 10 * 50 msec)

> IIUC we can only redo read logic for DOC_BUSY ERROR ?

Yes, that's to avoid having to penalize simple cache miss scenarios (i.e when there's no concurrent request). The new change uses write_lock_failure as an additional signal to identify the DOC_BUSY scenario. See below for more context on this.

https://cwiki.apache.org/confluence/display/TS/Presentations+-+2019?preview=/112821251/132320653/Collapsed%20Forwarding%20.pdf

File: proxy/http/HttpSM.cc (hunk 1 above)
Comment (sudheerv) [reply proxy/http/HttpSM.cc]: Yes, that's correct. That's to identify a doc_busy indirectly caused by a concurrent request to the same cache (using write lock failure signal).

**Comment by scw00** [comment]
@sudheerv Thanks for your response. Looks great.

**Comment by scw00** [comment]
```
Thread 34 "[ET_NET 31]" received signal SIGABRT, Aborted.
0x00007ffff593d207 in raise () from /lib64/libc.so.6
(gdb) r
The program being debugged has been started already.
Start it from the beginning? (y or n) y^CQuit
(gdb) bt
#0  0x00007ffff593d207 in raise () from /lib64/libc.so.6
#1  0x00007ffff593e8f8 in abort () from /lib64/libc.so.6
#2  0x00007ffff7aff0da in ink_abort (message_format=0x7ffff7b6c688 "%s:%d: failed assertion `%s`") at ink_error.cc:99
#3  0x00007ffff7afc7d2 in _ink_assert (expression=0xa49b10 "open_write_tries <= master_sm->t_state.txn_conf->max_cache_open_write_retries", file=0xa49868 "HttpCacheSM.cc", line=205) at ink_assert.cc:37
#4  0x00000000007a3bc0 in HttpCacheSM::state_cache_open_write (this=0x7fff7828e630, event=2, data=0x1146000) at HttpCacheSM.cc:205
#5  0x00000000006b05df in Continuation::handleEvent (this=0x7fff7828e630, event=2, data=0x1146000) at /home/scw00/trafficserver/iocore/eventsystem/I_Continuation.h:190
#6  0x000000000099dc27 in EThread::process_event (this=0x32f6810, e=0x1146000, calling_code=2) at UnixEThread.cc:136
#7  0x000000000099e202 in EThread::execute_regular (this=0x32f6810) at UnixEThread.cc:249
#8  0x000000000099e52b in EThread::execute (this=0x32f6810) at UnixEThread.cc:338
#9  0x000000000099cf1d in spawn_thread_internal (a=0x6578010) at Thread.cc:92
#10 0x00007ffff66fedd5 in start_thread () from /lib64/libpthread.so.0
#11 0x00007ffff5a04ead in clone () from /lib64/libc.so.6
(gdb) up
(gdb) p master_sm->t_state.txn_conf->max_cache_open_write_retries
$1 = 1
(gdb) p open_write_tries
$2 = 2
(gdb)

```

Abort when

File: proxy/http/HttpCacheSM.cc [hunk 3]
```diff
@@ -171,7 +172,18 @@ HttpCacheSM::state_cache_open_write(int event, void *data)
...
-    if (open_write_tries <= master_sm->t_state.txn_conf->max_cache_open_write_retries) {
+    if (master_sm->t_state.txn_conf->cache_open_write_fail_action == HttpTransact::CACHE_WL_FAIL_ACTION_READ_RETRY) {
+      // fall back to open_read_tries
+      if (open_write_tries <= master_sm->t_state.txn_conf->max_cache_open_write_retries) {
+        Debug("http_cache", "[%" PRId64 "] [state_cache_open_write] cache open write failure %d. read retry triggered",
+              master_sm->sm_id, open_write_tries);
+        open_read_tries          = 0;
+        read_retry_on_write_fail = true;
+        // make sure it doesn't loop indefinitely
+        open_write_tries = master_sm->t_state.txn_conf->max_cache_open_write_retries + 1;
```
Comment (scw00) [inline proxy/http/HttpCacheSM.cc]: should it be `open_write_tries ++` ?

**Comment by sudheerv** [comment]
> ```
> Thread 34 "[ET_NET 31]" received signal SIGABRT, Aborted.
> 0x00007ffff593d207 in raise () from /lib64/libc.so.6
> (gdb) r
> The program being debugged has been started already.
> Start it from the beginning? (y or n) y^CQuit
> (gdb) bt
> #0  0x00007ffff593d207 in raise () from /lib64/libc.so.6
> #1  0x00007ffff593e8f8 in abort () from /lib64/libc.so.6
> #2  0x00007ffff7aff0da in ink_abort (message_format=0x7ffff7b6c688 "%s:%d: failed assertion `%s`") at ink_error.cc:99
> #3  0x00007ffff7afc7d2 in _ink_assert (expression=0xa49b10 "open_write_tries <= master_sm->t_state.txn_conf->max_cache_open_write_retries", file=0xa49868 "HttpCacheSM.cc", line=205) at ink_assert.cc:37
> #4  0x00000000007a3bc0 in HttpCacheSM::state_cache_open_write (this=0x7fff7828e630, event=2, data=0x1146000) at HttpCacheSM.cc:205
> #5  0x00000000006b05df in Continuation::handleEvent (this=0x7fff7828e630, event=2, data=0x1146000) at /home/scw00/trafficserver/iocore/eventsystem/I_Continuation.h:190
> #6  0x000000000099dc27 in EThread::process_event (this=0x32f6810, e=0x1146000, calling_code=2) at UnixEThread.cc:136
> #7  0x000000000099e202 in EThread::execute_regular (this=0x32f6810) at UnixEThread.cc:249
> #8  0x000000000099e52b in EThread::execute (this=0x32f6810) at UnixEThread.cc:338
> #9  0x000000000099cf1d in spawn_thread_internal (a=0x6578010) at Thread.cc:92
> #10 0x00007ffff66fedd5 in start_thread () from /lib64/libpthread.so.0
> #11 0x00007ffff5a04ead in clone () from /lib64/libc.so.6
> (gdb) up
> (gdb) p master_sm->t_state.txn_conf->max_cache_open_write_retries
> $1 = 1
> (gdb) p open_write_tries
> $2 = 2
> (gdb)
> ```

Ah, that assert needs to be moved down. Thanks for testing and catching it!

File: proxy/http/HttpTransact.cc (hunk 2 above)
Comment (sudheerv) [reply proxy/http/HttpTransact.cc]: Done

**Comment by sudheerv** [comment]
> > ```
> > Thread 34 "[ET_NET 31]" received signal SIGABRT, Aborted.
> > 0x00007ffff593d207 in raise () from /lib64/libc.so.6
> > (gdb) r
> > The program being debugged has been started already.
> > Start it from the beginning? (y or n) y^CQuit
> > (gdb) bt
> > #0  0x00007ffff593d207 in raise () from /lib64/libc.so.6
> > #1  0x00007ffff593e8f8 in abort () from /lib64/libc.so.6
> > #2  0x00007ffff7aff0da in ink_abort (message_format=0x7ffff7b6c688 "%s:%d: failed assertion `%s`") at ink_error.cc:99
> > #3  0x00007ffff7afc7d2 in _ink_assert (expression=0xa49b10 "open_write_tries <= master_sm->t_state.txn_conf->max_cache_open_write_retries", file=0xa49868 "HttpCacheSM.cc", line=205) at ink_assert.cc:37
> > #4  0x00000000007a3bc0 in HttpCacheSM::state_cache_open_write (this=0x7fff7828e630, event=2, data=0x1146000) at HttpCacheSM.cc:205
> > #5  0x00000000006b05df in Continuation::handleEvent (this=0x7fff7828e630, event=2, data=0x1146000) at /home/scw00/trafficserver/iocore/eventsystem/I_Continuation.h:190
> > #6  0x000000000099dc27 in EThread::process_event (this=0x32f6810, e=0x1146000, calling_code=2) at UnixEThread.cc:136
> > #7  0x000000000099e202 in EThread::execute_regular (this=0x32f6810) at UnixEThread.cc:249
> > #8  0x000000000099e52b in EThread::execute (this=0x32f6810) at UnixEThread.cc:338
> > #9  0x000000000099cf1d in spawn_thread_internal (a=0x6578010) at Thread.cc:92
> > #10 0x00007ffff66fedd5 in start_thread () from /lib64/libpthread.so.0
> > #11 0x00007ffff5a04ead in clone () from /lib64/libc.so.6
> > (gdb) up
> > (gdb) p master_sm->t_state.txn_conf->max_cache_open_write_retries
> > $1 = 1
> > (gdb) p open_write_tries
> > $2 = 2
> > (gdb)
> > ```
> 
> 
> Ah, that assert needs to be moved down. Thanks for testing and catching it!

Fixed it

**Comment by sudheerv** [comment]
> Please update:
> 
> https://github.com/apache/trafficserver/blob/12afa068a44a9400392256fdc64d16a891c326b5/proxy/http/HttpTransact.cc#L2906-L2909

Done

**Comment by sudheerv** [comment]
> Please update:
> 
> https://github.com/apache/trafficserver/blob/12afa068a44a9400392256fdc64d16a891c326b5/proxy/http/HttpTransact.cc#L2906-L2909

**Comment by sudheerv** [comment]
> > Please update:
> > https://github.com/apache/trafficserver/blob/12afa068a44a9400392256fdc64d16a891c326b5/proxy/http/HttpTransact.cc#L2906-L2909
> 
> Done

**Comment by scw00** [comment]
```
Fatal: HttpSM.cc:2460: failed assertion `t_state.cache_open_write_fail_action == HttpTransact::CACHE_WL_FAIL_ACTION_READ_RETRY`

Thread 34 "[ET_NET 31]" received signal SIGABRT, Aborted.
[Switching to Thread 0x65b1700 (LWP 139755)]
0x00007ffff593d207 in raise () from /lib64/libc.so.6
Missing separate debuginfos, use: debuginfo-install bzip2-libs-1.0.6-13.el7.x86_64 elfutils-libelf-0.176-2.el7.x86_64 elfutils-libs-0.176-2.el7.x86_64 libattr-2.4.46-13.el7.x86_64 libcap-2.22-9.el7.x86_64 libgcc-4.8.5-36.el7.x86_64 libstdc++-4.8.5-36.el7.x86_64 pcre-8.32-17.el7.x86_64 sssd-client-1.16.2-13.el7_6.5.x86_64 systemd-libs-219-62.el7_6.5.x86_64 xz-libs-5.2.2-1.el7.x86_64 zlib-1.2.7-18.el7.x86_64
(gdb) bt
#0  0x00007ffff593d207 in raise () from /lib64/libc.so.6
#1  0x00007ffff593e8f8 in abort () from /lib64/libc.so.6
#2  0x00007ffff7aff0da in ink_abort (message_format=0x7ffff7b6c688 "%s:%d: failed assertion `%s`") at ink_error.cc:99
#3  0x00007ffff7afc7d2 in _ink_assert (expression=0xa63208 "t_state.cache_open_write_fail_action == HttpTransact::CACHE_WL_FAIL_ACTION_READ_RETRY", file=0xa61c52 "HttpSM.cc", line=2460)
    at ink_assert.cc:37
#4  0x000000000075d5ee in HttpSM::state_cache_open_write (this=0x7fff7828d0a0, event=1102, data=0x117ffa0) at HttpSM.cc:2460
#5  0x000000000075e1e4 in HttpSM::main_handler (this=0x7fff7828d0a0, event=1102, data=0x117ffa0) at HttpSM.cc:2614
#6  0x00000000006cddaf in Continuation::handleEvent (this=0x7fff7828d0a0, event=1102, data=0x117ffa0) at /home/scw00/trafficserver/iocore/eventsystem/I_Continuation.h:190
#7  0x00000000007c357b in HttpCacheSM::state_cache_open_write (this=0x7fff7828ee10, event=2, data=0x117ffa0) at HttpCacheSM.cc:209
#8  0x00000000006cddaf in Continuation::handleEvent (this=0x7fff7828ee10, event=2, data=0x117ffa0) at /home/scw00/trafficserver/iocore/eventsystem/I_Continuation.h:190
#9  0x00000000009bf4f1 in EThread::process_event (this=0x3330430, e=0x117ffa0, calling_code=2) at UnixEThread.cc:136
#10 0x00000000009bfacc in EThread::execute_regular (this=0x3330430) at UnixEThread.cc:249
#11 0x00000000009bfdf5 in EThread::execute (this=0x3330430) at UnixEThread.cc:338
#12 0x00000000009be7e7 in spawn_thread_internal (a=0x65b2010) at Thread.cc:92
#13 0x00007ffff66fedd5 in start_thread () from /lib64/libpthread.so.0
#14 0x00007ffff5a04ead in clone () from /lib64/libc.so.6

```

**Comment by sudheerv** [comment]
> che_open_wri

Fixed the assert. Thanks @scw00 for catching it!

File: proxy/http/HttpCacheSM.cc [hunk 4]
```diff
@@ -159,7 +159,8 @@ HttpCacheSM::state_cache_open_write(int event, void *data)
 {
   STATE_ENTER(&HttpCacheSM::state_cache_open_write, event);
   ink_assert(captive_action.cancelled == 0);
```
Comment (scw00) [inline proxy/http/HttpCacheSM.cc]: Please reinit the captive_action. Otherwise this action will be dirty by `pending_action->cancel` in HttpSM.

**Comment by scw00** [comment]
I have some concerns :
 - It involves COMPLETE_LOOK_UP_HOOK twice might leading some memleaking in plugins. Please update doc if it is expected.
 - What should we do for update scenario. It seems we do return wrong object directly once we failed to get vol->lock (without retry).

File: proxy/http/HttpCacheSM.cc [hunk 5]
```diff
@@ -171,7 +172,18 @@ HttpCacheSM::state_cache_open_write(int event, void *data)
     break;
 
   case CACHE_EVENT_OPEN_WRITE_FAILED:
-    if (open_write_tries <= master_sm->t_state.txn_conf->max_cache_open_write_retries) {
+    if (master_sm->t_state.txn_conf->cache_open_write_fail_action == HttpTransact::CACHE_WL_FAIL_ACTION_READ_RETRY) {
+      // fall back to open_read_tries
+      if (open_write_tries <= master_sm->t_state.txn_conf->max_cache_open_write_retries) {
+        Debug("http_cache", "[%" PRId64 "] [state_cache_open_write] cache open write failure %d. read retry triggered",
```
Comment (scw00) [inline proxy/http/HttpCacheSM.cc]: What if `object_info != nullptr`. Does it disable the write retries even if the update case.

File: proxy/http/HttpCacheSM.cc (hunk 4 above)
Comment (sudheerv) [reply proxy/http/HttpCacheSM.cc]: Fixed. Thanks

File: proxy/http/HttpCacheSM.cc (hunk 5 above)
Comment (sudheerv) [reply proxy/http/HttpCacheSM.cc]: Updated the docs and created a new PR. Closing this one. Thanks for the review!

**Comment by zwoop** [comment]
@sudheerv  Are you making another PR for this? :)

**Pull request closed without merging**